pkg/handler/list_owner_nft.go

```diff
@@ -22,19 +31,143 @@ func ConfigureListOwnerNFTRoute(route httproute.Route) httproute.Route {
 
 type ListOwnerNFTHandlerLogger struct{ *log.Logger }
 
+type ListOwnerNFTHandlerNFTOwnerQuery interface {
+	QueryOwner(ownerID authgearweb3.ContractID) (*dbmodel.NFTOwner, error)
+}
+
+type ListOwnerNFTHandlerNFTOwnershipMutator interface {
+	InsertNFTOwnerships(ownerships []dbmodel.NFTOwnership) error
+}
+
+type ListOwnerNFTHandlerAlchemyAPI interface {
+	GetOwnerNFTs(ownerAddress string, contractIDs []authgearweb3.ContractID, pageKey string) (*apimodel.GetNFTsResponse, error)
+	GetAssetTransfers(contractIDs []authgearweb3.ContractID, fromAddress string, toAddress string, fromBlock string, toBlock string, pageKey string, maxCount int64, order string) (*apimodel.AssetTransferResponse, error)
+}
+
 func NewListOwnerNFTHandlerLogger(lf *log.Factory) ListOwnerNFTHandlerLogger {
 	return ListOwnerNFTHandlerLogger{lf.New("api-list-owner-nft")}
 }
 
 type ListOwnerNFTAPIHandler struct {
-	JSON               JSONResponseWriter
-	Logger             ListOwnerNFTHandlerLogger
-	Config             config.Config
-	NFTOwnerQuery      query.NFTOwnerQuery
-	NFTCollectionQuery query.NFTCollectionQuery
+	JSON                JSONResponseWriter
+	Logger              ListOwnerNFTHandlerLogger
+	Config              config.Config
+	AlchemyAPI          ListOwnerNFTHandlerAlchemyAPI
+	NFTOwnerQuery       ListOwnerNFTHandlerNFTOwnerQuery
+	NFTCollectionQuery  query.NFTCollectionQuery
+	NFTOwnershipQuery   query.NFTOwnershipQuery
+	NFTOwnershipMutator ListOwnerNFTHandlerNFTOwnershipMutator
+}
+
+func (h *ListOwnerNFTAPIHandler) FetchAndInsertNFTTransfers(ownerID authgearweb3.ContractID, contracts []authgearweb3.ContractID) error {
+	ownerships := make([]database.NFTOwnership, 0)
+	pageKey := ""
+	nftFetchCount := 0
+	contractIDsToEnquire := make([]authgearweb3.ContractID, 0)
+	contractTokenIDToBalance := make(map[ContractTokenID]string)
+	// Fetch user nfts until no extra page or has reached the page limit
+	for ok := true; ok; ok = pageKey != "" && nftFetchCount <= h.Config.Server.MaxNFTPages {
+		nfts, err := h.AlchemyAPI.GetOwnerNFTs(ownerID.ContractAddress, contracts, pageKey)
+		if err != nil {
+			return err
+		}
+
+		for _, ownedNFT := range nfts.OwnedNFTs {
+
+			contractID := authgearweb3.ContractID{
+				Blockchain:      ownerID.Blockchain,
+				Network:         ownerID.Network,
+				ContractAddress: ownedNFT.Contract.Address,
```

Comment:
This is unsafe. The address may be mixed case. Add a NewContractID function to construct a safe contractID.

In general, parsing from alchemy model to our own model WITHOUT a dedicated parse function is unsafe. Introduce a parse function for each possible parsing step instead of constructing a model directly.

Reply:
I have opened https://github.com/authgear/authgear-server/pull/2551 for the NewContractID